dotnet/src/SemanticKernel/Reliability/DefaultHttpRetryHandler.cs

```diff
@@ -0,0 +1,259 @@
+        {
+            newRequest.Headers.TryAddWithoutValidation(header.Key, header.Value);
+        }
+
+        // Copy request properties.
+        foreach (var property in originalRequest.Properties)
+        {
+            newRequest.Properties.Add(property);
+        }
+
+        // Set Content if previous request had one.
+        if (originalRequest.Content != null)
+        {
+            // HttpClient doesn't rewind streams and we have to explicitly do so.
+            await originalRequest.Content.ReadAsStreamAsync().ContinueWith(t =>
+            {
+                if (t.Result.CanSeek) { t.Result.Seek(0, SeekOrigin.Begin); }
+
+                newRequest.Content = new StreamContent(t.Result);
+            }, TaskScheduler.Current).ConfigureAwait(false);
+
+            // Copy content headers.
+            if (originalRequest.Content.Headers != null)
+            {
+                foreach (var contentHeader in originalRequest.Content.Headers)
+                {
+                    newRequest.Content.Headers.TryAddWithoutValidation(contentHeader.Key, contentHeader.Value);
+                }
+            }
+        }
+
+        return newRequest;
+    }
+
+    /// <summary>
+    /// Executes the action with retry logic
+    /// </summary>
+    /// <remarks>
+    /// The request is retried if it throws an exception that is a retryable exception.
+    /// If the request throws an exception that is not a retryable exception, it is not retried.
+    /// If the request returns a response with a retryable error code, it is retried.
+    /// If the request returns a response with a non-retryable error code, it is not retried.
+    /// If the exception contains a RetryAfter header, the request is retried after the specified delay.
+    /// If configured to use exponential backoff, the delay is doubled for each retry.
+    /// </remarks>
+    /// <param name="request">The request.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        int retryCount = 0;
+
+        var start = this._timeProvider.GetCurrentTime();
+        while (true)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            TimeSpan waitFor = default;
+            string reason = string.Empty;
+            HttpResponseMessage? response = null;
+            try
+            {
+                response = await base.SendAsync(request, cancellationToken);
+
+                // If the request does not require a retry then we're done
+                if (!this.ShouldRetry(response.StatusCode))
+                {
+                    return response;
+                }
+
+                // Drain response content to free connections. Need to perform this
+                // before retry attempt and before the TooManyRetries ServiceException.
```

Comment:
This was suggested in the initial PR and I'm finding other examples like here in OneDrive https://github.com/OneDrive/samples/blob/master/scenarios/throttling-ratelimit-handling/RateLimitDemo/ThrottlingHandler.cs

I've also found note in msgraph-sdk-dotnet-core about doing this, however this notes only when custom handler is throwing an exception, so not clear if applicable in our scenario. Will do some more investigation.
https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/blob/dev/docs/logging-requests.md

Reply:
Yeah, for now I'll go ahead and remove, since we aren't throwing an exception in this path, we return the request, it's on the caller to do this which I think is the right thing to do.

Reply:
Actually, I tried some tests and confirmed there are left over threads/connections that don't get cleaned up.

Without drain:
![image](https://user-images.githubusercontent.com/35783676/224791061-593b763d-0f38-4762-9db1-b146ba4dd691.png)

With drain:
![image](https://user-images.githubusercontent.com/35783676/224791119-668a742e-1308-4e50-a54c-918766c74d7f.png)

Reply:
Resolving as-is.

Reply:
this is where I'm starting to think we don't exactly know how it works and we're guessing, risking of making things worse. I hate keeping this PR hanging, but I think we either master this code or we take a dependency on something battle tested. 

For instance, leaving aside MSGraph for a second, how do ASP.NET and Polly handle this problem?

Thoughts?

Reply:
I'm not sure, I'll continue to try and find an answer.

I can remove the clone and replace the ReadContent with a dispose as the end of the loop to avoid the negative behavior above.

Reply:
As far as I understand, Polly doesn't implement an http message handler as their policies can be applied more generically. https://github.com/App-vNext/Polly/blob/7903c9578f6dc14241c747818cf4c9e1c62aa8d1/src/Polly/Retry/RetryEngine.cs